### PR TITLE
Centralized logging with Kinesis Firehose using Lambda Extensions

### PR DESCRIPTION
**Pull request by hariohmprasath**
*Issue #, if available:*
Lambda Extension to send logs directly to Kinesis firehose #51 

*Description of changes:*
The extension explains the overall approach to **streamline and and centralize log collection** using Kinesis firehose. The extension runs a local HTTP listener and subscribes to a stream of function and platform logs using the Logs API. It buffers them and sends them to Kinesis firehose periodically. The demo gets deployed using AWS SAM.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: kinesisfirehose-logs-extension-demo/README.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+* AWS SAM CLI needs to get installed, follow the [link](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) to learn how to install them
+
+### Build
+
+Check out the code by running the following command:
+
+```bash
+mkdir kinesisfirehose-logs-extension-demo && cd kinesisfirehose-logs-extension-demo
+git clone https://github.com/hariohmprasath/load-testing-serverless-apps.git .
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Keep this simpler by cloning the whole repo and then cd into your subdirectory

File: README.md [hunk 2]
```diff
@@ -55,6 +55,8 @@ In this repository you will find a number of demos and sample projects from AWS
 
 * [Demo: Logs to Amazon S3 extension: container image ](s3-logs-extension-demo-container-image/): Demo logs extension to receive logs directly from Lambda and send them to S3. This example packages the extension and function as separate container images. The demo is deployed using AWS SAM.
 
+* [Demo: Logs to Kinesis firehose Logs API extension](kinesisfirehose-logs-extension-demo/): How to get a basic logs API extension for AWS Kinesis Firehose written in Go. The extension explains the overall approach to streamline and and centralize log collection using Kinesis firehose. The extension runs a local HTTP listener and subscribes to a stream of function and platform logs using the Logs API. It buffers them and sends them to Kinesis firehose periodically. The demo gets deployed using AWS SAM.
```
Comment (julianwood) [inline README.md]: Ensure first reference is Amazon Kinesis Data Firehose, then you can refer to Kinesis or Kinesis Data Firehose afterwards.

Comment (julianwood) [reply README.md]: Also worth mentioning that Kinesis sends the logs to S3

File: kinesisfirehose-logs-extension-demo/README.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
+# Centralize log collection with Kinesis Firehose using Lambda Extensions
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Amazon Kinesis Data Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+## Introduction
+
+This pattern walks through an approach to centralize log collection for lambda function with Kinesis firehose using external extensions. The provided code sample shows how to get send logs directly to kinesis firehose without sending them to AWS CloudWatch service.
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Always Lambda....all AWS/Amazon products have capital letters so check these for Lambda/Kinesis/etc.

File: kinesisfirehose-logs-extension-demo/README.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+> Note: This is a simple example extension to help you investigate an approach to centralize the log aggregation. This example code is not production ready. Use it with your own discretion after testing thoroughly.
+
+This sample extension:
+
+* Subscribes to receive `platform` and `function` logs.
+* Runs with a main, and a helper goroutine: The main goroutine registers to `ExtensionAPI` and process its `invoke` and `shutdown` events. The helper goroutine:
+  * starts a local HTTP server at the provided port (default 1234) that receives requests from Logs API with `NextEvent` method call
+  * puts the logs in a synchronized queue (Producer) to be processed by the main goroutine (Consumer)
+* The main goroutine writes the received logs to AWS Kinesis firehose, which gets stored in AWS S3
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Amazon Kinesis and Amazon S3

File: kinesisfirehose-logs-extension-demo/README.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+* fetching configuration settings or secrets before the function invocation
+* detecting and alerting on function activity through hardened security agents, which can run as separate processes from the function
+
+read more about it [here](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-extensions-in-preview/)
+
+> Note: The code sample provided part of this pattern uses **external** extension to listen to log events from the lambda function
+
+## Need to centralize log collection
+
+Having a centralized log collection mechanism using kinesis firehose provides the following benefits:
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Kinesis Data Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+* Define and standardize the transformations before the log gets delivered to downstream systems like S3, elastic search, redshift, etc
+* Provides a secure storage area for log data, before it gets written out to the disk. In the event of machine/application failure, we still have access to the logs emitted from the source machine/application
+
+## Architecture
+
+### AWS Services
+
+* AWS Lambda
+* AWS Lambda extension
+* AWS KinesisFirehose
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Amazon Kinesis Data Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+* AWS S3
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Amazon S3

File: kinesisfirehose-logs-extension-demo/README.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+Once deployed the overall flow looks like below:
+
+* On start-up, the extension subscribes to receive logs for `Platform` and `Function` events.
+* A local HTTP server is started inside the external extension which receives the logs.
+* The extension also takes care of buffering the recieved log events in a synchronized queue and writing it to AWS Kinesis Firehose via direct `PUT` records
+
+> Note: Firehose stream name gets specified as an environment variable (`AWS_KINESIS_STREAM_NAME`)
+
+* The lambda function won't be able to send any logs events to AWS CloudWatch service due to the following explict `DENY` policy:
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Amazon CloudWatch

File: kinesisfirehose-logs-extension-demo/README.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+```bash
+{
+    "StatusCode": 200,
+    "LogResult": "<<Encoded>>",
+    "ExecutedVersion": "$LATEST"
+}
+```
+
+In a few minutes after the successfully invocation of the lambda function, we should start seeing the log messages from the example extension written to an S3 bucket.
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: I think its worth expanding this slightly to say that the log messages are sent to Kinesis Data Firehose which sends the messages to S3.

File: kinesisfirehose-logs-extension-demo/agent/http.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+	"os"
+	"time"
+
+	"aws-lambda-extensions/kinesisfirehose-logs-extension-demo/logsapi"
+
+	"github.com/golang-collections/go-datastructures/queue"
+)
+
+// DefaultHttpListenerPort is used to set the URL where the logs will be sent by Logs API
+const DefaultHttpListenerPort = "1234"
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/agent/http.go]: Is it worth having this as a default which can be overwritten by an ENVAR?

File: kinesisfirehose-logs-extension-demo/agent/http.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+		MaxItems:  10000,
+		MaxBytes:  262144,
+		TimeoutMS: 1000,
+	}
+	if err != nil {
+		return err
+	}
+	destination := logsapi.Destination{
+		Protocol:   logsapi.HttpProto,
+		URI:        logsapi.URI(fmt.Sprintf("http://sandbox:%s", DefaultHttpListenerPort)),
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/agent/http.go]: Best practice is to use `sandbox.localdomain`

**Review by julianwood — CHANGES_REQUESTED** [review]
It may also be worth thinking about sending events per invoke. You can use the Platform `runtimeDone` messages to in effect wait until all the logs have arrived from Lambda before calling /next. This doesn't slow down the function response and makes for cleaner logs when the execution environment is frozen between invocations which may mean only receiving the logs during the next invoke.

File: kinesisfirehose-logs-extension-demo/README.md (hunk 1 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated with instructions to clone the whole repository

File: README.md (hunk 2 above)
Comment (hariohmprasath) [reply README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 4 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 5 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 6 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 8 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 9 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 10 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/agent/http.go (hunk 11 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/agent/http.go]: Added new environment variable `HTTP_LOGS_LISTENER_PORT` which allows customization of the default http ports for receiving log events

File: kinesisfirehose-logs-extension-demo/agent/http.go (hunk 12 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/agent/http.go]: Updated

**Comment by hariohmprasath** [comment]
> It may also be worth thinking about sending events per invoke. You can use the Platform `runtimeDone` messages to in effect wait until all the logs have arrived from Lambda before calling /next. This doesn't slow down the function response and makes for cleaner logs when the execution environment is frozen between invocations which may mean only receiving the logs during the next invoke.

Yep, 
The code flushes logs only incase of `Shutdown` and `runtimeDone` events. Here is a snippet of it.

`
flushLogQueue := func(force bool) {
		for !(logQueue.Empty() && (force || strings.Contains(logsStr, string(logsapi.RuntimeDone)))) {
           // Flush logs
      }
}
`

`
flushLogQueue(true/false)
`

File: kinesisfirehose-logs-extension-demo/README.md (hunk 7 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md (hunk 3 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Updated

File: kinesisfirehose-logs-extension-demo/README.md [hunk 13]
```diff
@@ -0,0 +1,226 @@
+# Centralize log collection with Amazon Kinesis Data Firehose using Lambda Extensions
+
+## Introduction
+
+This pattern walks through an approach to centralize log collection for Lambda function with Kinesis firehose using external extensions. The provided code sample shows how to get send logs directly to Kinesis firehose without sending them to AWS CloudWatch service.
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Suggestion: This example show how to centralize log collection for a Lambda function using Kinesis Data Firehose. The provided code sample uses Lambda extensions to receive logs from Lambda and send them directly to Kinesis Data Firehose without sending them to Amazon CloudWatch service.

File: kinesisfirehose-logs-extension-demo/README.md [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+> Note: This is a simple example extension to help you investigate an approach to centralize the log aggregation. This example code is not production ready. Use it with your own discretion after testing thoroughly.
+
+This sample extension:
+
+* Subscribes to receive `platform` and `function` logs.
+* Runs with a main, and a helper goroutine: The main goroutine registers to `ExtensionAPI` and process its `invoke` and `shutdown` events. The helper goroutine:
+  * starts a local HTTP server at the provided port (default 1234, the port can be overridden with Lambda environment variable `HTTP_LOGS_LISTENER_PORT` ) that receives requests from Logs API with `NextEvent` method call
+  * puts the logs in a synchronized queue (Producer) to be processed by the main goroutine (Consumer)
+* The main goroutine writes the received logs to Amazon Kinesis firehose, which gets stored in Amazon S3
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Amazon Kinesis Data Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+## Amazon Kinesis Data firehose
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md:17]: Firehose (perhaps add service page link: https://aws.amazon.com/kinesis/data-firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+Amazon Kinesis Data Firehose is the easiest way to reliably load streaming data into data lakes, data stores, and analytics services. It can capture, transform, and deliver streaming data to Amazon S3, Amazon Redshift, Amazon Elasticsearch Service, generic HTTP endpoints, and service providers like Datadog, New Relic, MongoDB, and Splunk, read more about it [here](https://aws.amazon.com/kinesis/data-firehose)
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md:19]: Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+* fetching configuration settings or secrets before the function invocation
+* detecting and alerting on function activity through hardened security agents, which can run as separate processes from the function
+
+read more about it [here](https://aws.amazon.com/blogs/compute/introducing-aws-lambda-extensions-in-preview/)
+
+> Note: The code sample provided part of this pattern uses **external** extension to listen to log events from the Lambda function
+
+## Need to centralize log collection
+
+Having a centralized log collecting mechanism using Kinesis firehose provides the following benefits:
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Kinesis Data Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+
+* Helps to collect logs from different sources in one place. Even though the sample provided sends logs from Lambda, log routers like `Fluentbit` and `Firelens` can send logs directly to Kinesis Data firehose from container orchestrators like `EKS` and `ECS`.
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Kinesis Data Firehose

Comment (julianwood) [reply kinesisfirehose-logs-extension-demo/README.md]: Expand the service names on 1st mention so ppl know what they refer to [Amazon Elastic Kubernetes Service (EKS)]()" and "[Amazon Elastic Container Service (ECS)](https://aws.amazon.com/eks)"

File: kinesisfirehose-logs-extension-demo/README.md [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+* Define and standardize the transformations before the log gets delivered to downstream systems like S3, elastic search, redshift, etc
+* Provides a secure storage area for log data, before it gets written out to the disk. In the event of machine/application failure, we still have access to the logs emitted from the source machine/application
+
+## Architecture
+
+### AWS Services
+
+* Amazon Lambda
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: AWS Lambda

File: kinesisfirehose-logs-extension-demo/README.md [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+* Amazon Lambda extension
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: AWS Lambda

File: kinesisfirehose-logs-extension-demo/README.md [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+Sid: CloudWatchLogsDeny
+Effect: Deny
+Action:
+  - logs:CreateLogGroup
+  - logs:CreateLogStream
+  - logs:PutLogEvents
+Resource: arn:aws:logs:*:*:*
+```
+
+* The Kinesis Firehose stream configured part of this sample sends log directly to `AWS S3` (gzip compressed).
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Kinesis Data Firehose

File: kinesisfirehose-logs-extension-demo/README.md [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+
+## Build and Deploy
+
+AWS SAM template available part of the root directory can be used for deploying the sample lambda function with this extension
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Lambda

File: kinesisfirehose-logs-extension-demo/README.md [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+### Pre-requistes
+
+* AWS SAM CLI needs to get installed, follow the [link](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) to learn how to install them
+
+### Build
+
+Check out the code by running the following command:
+
+```bash
+mkdir aws-lambda-extensions && cd aws-lambda-extensions
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Do you need to make the directory, the clone will create it anyway?

File: kinesisfirehose-logs-extension-demo/README.md [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+
+Key                 HelloWorldFunction
+Description         First Lambda Function ARN
+Value               arn:aws:lambda:us-east-1:xxx:function:kinesisfirehose-logs-extension-demo-function
+-------------------------------------------------------------------------------------------------------------------
+```
+
+## Testing
+
+You can invoke the Lambda function using the following CLI command
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: Suggestion: "You can invoke the Lambda function using the [Lambda Console](https://console.aws.amazon.com/lambda/home), or the following CLI command"

File: kinesisfirehose-logs-extension-demo/README.md [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+
+```bash
+{
+    "StatusCode": 200,
+    "LogResult": "<<Encoded>>",
+    "ExecutedVersion": "$LATEST"
+}
+```
+
+In a few minutes after the successful invocation of the Lambda function, we should start seeing the log messages from the example extension sent to Amazon Data Firehose which sends the messages to a Amazon S3 bucket.
```
Comment (julianwood) [inline kinesisfirehose-logs-extension-demo/README.md]: suggestion "you should see the"

Comment (julianwood) [reply kinesisfirehose-logs-extension-demo/README.md]: Kinesis Data Firehose

**Review by julianwood — CHANGES_REQUESTED** [review]
Looking good, just some working suggestions in the README for clarity and correct use of service names.

File: kinesisfirehose-logs-extension-demo/README.md (hunk 23 above)
Comment (hariohmprasath) [reply kinesisfirehose-logs-extension-demo/README.md]: Removed extra mkdir command

**Comment by hariohmprasath** [comment]
> Looking good, just some working suggestions in the README for clarity and correct use of service names.

Thanks @julianwood, all the changes are incorporated.

**Comment by julianwood** [comment]
Thanks for contributing.

**Pull request merged**